crates/revm_precompiles/src/fhe.rs

```diff
@@ -1,492 +1,93 @@
-use std;
-
-use super::{CustomPrecompileFn, Precompile, PrecompileResult, Return as Error};
+use super::{Precompile, PrecompileResult, Return};
 use crate::{Address, PrecompileOutput};
 
-use once_cell::sync::Lazy;
-use sunscreen::{
-    fhe_program,
-    types::{bfv::Signed, Cipher},
-    Application, Ciphertext, Compiler, FheProgramInput, PublicKey, RuntimeError,
-};
-use sunscreen_params::{SUNSCREEN_PARAMS, SUNSCREEN_RUNTIME};
+use fhe_precompiles::testnet::one::FHE;
+use fhe_precompiles::FheError;
 
 pub const COST_FHE_ADD: u64 = 200;
 pub const COST_FHE_ADD_PLAIN: u64 = 200;
 pub const COST_FHE_SUBTRACT: u64 = 200;
 pub const COST_FHE_SUBTRACT_PLAIN: u64 = 200;
 pub const COST_FHE_MULTIPLY: u64 = 1000;
-pub const COST_FHE_ENC_ZERO: u64 = 100;
+pub const COST_FHE_ENCRYPT_ZERO: u64 = 100;
 
-// TODO This should maybe go in a separate crate that the wallet imports as
-// well, to ensure the same params are used.
-pub static FHE_APP: Lazy<Application> = Lazy::new(|| {
-    Compiler::new()
-        .fhe_program(add)
-        .fhe_program(add_plain)
-        .fhe_program(subtract)
-        .fhe_program(subtract_plain)
-        .fhe_program(multiply)
-        .with_params(&SUNSCREEN_PARAMS)
-        .compile()
-        .unwrap()
-});
+impl From<FheError> for Return {
+    fn from(value: FheError) -> Self {
+        match value {
+            FheError::UnexpectedEOF => Return::Other("Not enough input".into()),
+            FheError::PlatformArchitecture => {
+                Return::Other("Validator needs at least 32B architecture".into())
+            }
+            FheError::InvalidEncoding => Return::Other("Invalid bincode encoding".into()),
+            FheError::Overflow => Return::Other("i64 overflow".into()),
+            FheError::SunscreenError(e) => {
+                Return::Other(format!("Sunscreen error: {:?}", e).into())
+            }
+        }
+    }
+}
+
+fn to_precompile<F>(f: F, input: &[u8], op_cost: u64, gas_limit: u64) -> PrecompileResult
+where
+    F: Fn(&[u8]) -> Result<Vec<u8>, FheError>,
+{
+    if op_cost > gas_limit {
+        return Err(Return::OutOfGas);
+    }
+
+    f(input)
+        .map(|x| PrecompileOutput::without_logs(op_cost, x))
+        .map_err(|e| e.into())
+}
 
-// For people making other contracts, allow passing in the program itself
-// Will require on the fly runtime
-//pub const FHE_ARBITRARY: (B160, Precompile) = (
-//u64_to_b160(204), // 0xcc
-//Precompile::Custom(fhe_arbitrary as CustomPrecompileFn),
-//);
 pub const FHE_ADD: (Address, Precompile) = (
     crate::make_address(0, 205), // 0xcd
-    Precompile::Custom(fhe_add as CustomPrecompileFn),
+    Precompile::Custom(|input, gas_limit| {
+        to_precompile(|x| FHE.add(x), input, COST_FHE_ADD, gas_limit)
```

Comment:
You could also make `to_precompile` return the `Precompile::Custom` itself if you want.